### PR TITLE
onedrive: upgrade to Graph with olihey/onedrive_graph

### PR DESCRIPTION
**Pull request by Cnly**
This is an attempt to merge olihey/onedrive_graph into master (with some other changes). The merged code is not compatible with current config files for OneDrive backends.

### CODE REVIEW

**Comment by Cnly** [comment]
> In what way isn't it compatible?

The main thing will be rclone's Client ID / Key. The old pair does not work for Graph (HTTP 401). Also, the new code keeps `driveType` and `graphID` (i.e. drive ID) in config, which are not present in the current config.

Can we somehow convert the token? If not, maybe re-doing the auth will be necessary.

> Losing the modtime is quite a big deal for rclone as the syncing won't work properly without it. Perhaps we should consider this idea in a separate PR?

Didn't think about that before... I've cherry-picked the commit that disabled single part upload here now.

**Comment by ncw** [comment]
> The main thing will be rclone's Client ID / Key. The old pair does not work for Graph (HTTP 401). Also, the new code keeps driveType and graphID (i.e. drive ID) in config, which are not present in the current config.

OK.

Can you make a message for the user which tells them they need to re-run `rclone config` if `driveType` and `graphID` aren't defined?  That would at least provide feedback for the user as to what they need to do.

I'll give the code a quick go through now.

Do you think this code will be ready for the 1.43 release?  Which is scheduled for 1st september?  I try to have a week of feature freeze before the release so I'm thinking it might be best to slip this until after then - what do you think?

File: backend/onedrive/onedrive.go [hunk 1]
```diff
@@ -33,48 +32,32 @@ import (
...
-	rootURLPersonal                     = "https://api.onedrive.com/v1.0/drive" // root URL for requests
-	discoveryServiceURL                 = "https://api.office.com/discovery/"
-	configResourceURL                   = "resource_url"
+	rcloneClientID              = "b15665d9-eda6-4092-8539-0eec376afd59"
+	rcloneEncryptedClientSecret = "_JUdzh3LnKNqSPcf4Wu5fgMFIQOI8glZu_akYgR8yf6egowNBg-R"
+	minSleep                    = 10 * time.Millisecond
+	maxSleep                    = 2 * time.Second
+	decayConstant               = 2 // bigger for slower decay, exponential
+	graphURL                    = "https://graph.microsoft.com/v1.0"
+	configGraphID               = "graphID"
```
Comment (ncw) [inline backend/onedrive/onedrive.go]: By convention the config file names should be `graph_id` and `drive_type`.

Anything stored in the config file should also have entries in RegInfo.Options (as advanced options) so they can be configured by environment variable or command line

File: backend/onedrive/onedrive.go [hunk 2]
```diff
@@ -84,147 +67,151 @@ func init() {
...
+			opts = rest.Opts{
+				Method:  "GET",
+				RootURL: graphURL,
+				Path:    "/drives/" + finalDriveID + "/root"}
+			var rootItem api.Item
+			resp, err := srv.CallJSON(&opts, nil, &rootItem)
+			if err != nil {
+				log.Fatalf("Failed to query root for drive %s: %v", finalDriveID, err)
+			}
+			if resp.StatusCode != 200 {
```
Comment (ncw) [inline backend/onedrive/onedrive.go]: You'll get an error from `CallJSON` if it returns > 299 response so this isn't necessary

File: backend/onedrive/onedrive.go [hunk 3]
```diff
@@ -84,147 +67,151 @@ func init() {
...
-
-				var foundService string
-				if len(resourcesID) == 0 {
-					fs.Errorf(nil, "No Service found")
-					return
-				} else if len(resourcesID) == 1 {
-					foundService = resourcesID[0]
-				} else {
-					foundService = config.Choose("Choose resource URL", resourcesID, resourcesURL, false)
+				if resp.StatusCode != 200 {
```
Comment (ncw) [inline backend/onedrive/onedrive.go]: You'll get an error from CallJSON if it returns > 299 response so this isn't necessary

File: backend/onedrive/onedrive.go [hunk 4]
```diff
@@ -84,147 +67,151 @@ func init() {
...
+			// query Microsoft Graph
+			if finalDriveID == "" {
+				drives := drivesResponse{}
+				resp, err := srv.CallJSON(&opts, nil, &drives)
 				if err != nil {
-					fs.Errorf(nil, "Failed to get resource token: %v", err)
-					return
+					log.Fatalf("Failed to query available drives: %v", err)
 				}
 				if resp.StatusCode != 200 {
```
Comment (ncw) [inline backend/onedrive/onedrive.go]: You'll get an error from CallJSON if it returns > 299 response so this isn't necessary

**Comment by ncw** [comment]
> Do you think this code will be ready for the 1.43 release? Which is scheduled for 1st september? I try to have a week of feature freeze before the release so I'm thinking it might be best to slip this until after then - what do you think?

Or if we can merge it in the next couple of days I can ask people to test it on the forum before the release...

**Comment by ncw** [comment]
I see I broke the build :-(  I'll fix that in a moment - can you rebase?

**Comment by Cnly** [comment]
@ncw I read the build log before, and it seems the two CI's both attempt to merge the pull request before building. Therefore other code changes in master break builds here too... Maybe that's out of my control?

**Comment by ncw** [comment]
@Cnly 
> I read the build log before, and it seems the two CI's both attempt to merge the pull request before building. Therefore other code changes in master break builds here too... Maybe that's out of my control?

I think if you rebase your branch on master and force push it here then it will fix it.

**Comment by Cnly** [comment]
Forgot to say that I've rebased...

Is there any facility, for example Travis, that can compile and provide rclone for all platforms for this pull request? That may make it easier to distribute these builds to the testers on the forum.

**Comment by ncw** [comment]
I re-kicked the CI and it built fine!

> Is there any facility, for example Travis, that can compile and provide rclone for all platforms for this pull request? That may make it easier to distribute these builds to the testers on the forum.

I've pushed this branch to the `Cnly-graph` on the rclone repository.  This means that the CI will build a beta - you'll find it here if you want to tell people about it

https://beta.rclone.org/branch/v1.42-133-gb874bc22-Cnly-graph/

I will try to find the time to test and review this evening.

**Comment by Cnly** [comment]
@ncw Thanks! I've posted this on the forum: https://forum.rclone.org/t/please-help-to-test-out-new-api-for-onedrive-less-throttling-better-sharepoint-support-etc/

**Comment by Cnly** [comment]
@ncw Do you think this branch can get into v1.43? Currently except for some long-standing problems, things are going okay.

The long-standing problems are:
- Sharepoint modifies files in some specific formats when they're uploaded
- Microsoft doesn't want us to navigate remote folders ("shared with me" in OneDrive Personal) by path. We'd have to know exactly the itemID. This has been partially solved in #1200, #2369. We're still unable to, for example, do `newFs` with folders inside a shared folder as root.

which seem to be more of Microsoft's problem...

**Comment by ncw** [comment]
I've been following your testing progress on the forum with interest.

I think what we should do is merge this immediately after 1.43 has been released.  I know that is disappointing for you but I really don't want to break the release at the last minute.  It has been partly my fault too as I haven't had enough time to test it myself.  The next release will be around in 6 weeks which is a nice length of time to get it stabilised and get the bugs worked out.

Would you like to be a maintainer of rclone, maybe specifically for the onedrive backend?  If so drop me an email to nick@craig-wood.com - thanks

**Comment by Cnly** [comment]
@ncw Okay, I would agree with you. It's nothing wrong - it would be better if we ensure this is stable enough. Maybe we can have time to look deeper into the problems too. :)

Regarding maintainer of rclone: Yes, I'll reach you through email in a short while.

**Comment by ncw** [comment]
OK now I've got 1.43 out of the door, I've been though this and I've merged it to master :-)

I've squashed a few of the commits and tidied up the attribution in the merge commit but otherwise the code is unchanged.

I like the message when you use an old onedrive remote. 

I think the upgrade procedure could work slightly better - if you start with an old remote like this

```
[onedrive]
type = onedrive
client_id = 
client_secret = 
token = {"access_token":"..."}
```

Then run `rclone config` then `e` to edit the remote, you get

```
$ rclone config 
Current remotes:

Name                 Type
====                 ====
[snip]

e) Edit existing remote
n) New remote
d) Delete remote
r) Rename remote
c) Copy remote
s) Set configuration password
q) Quit config
e/n/d/r/c/s/q> e
Choose a number from below, or type in an existing value
[snip]
remote> onedrive
--------------------
[onedrive]
type = onedrive
client_id = 
client_secret = 
token = {"access_token":"..."}
--------------------
Edit remote
Value "client_id" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Value "client_secret" = ""
Edit? (y/n)>
y) Yes
n) No
y/n> n
Edit advanced config? (y/n)
y) Yes
n) No
y/n> n
--------------------
[onedrive]
type = onedrive
client_id = 
client_secret = 
token = {"access_token":"..."}
--------------------
y) Yes this is OK
e) Edit this remote
d) Delete this remote
y/e/d> y
Remote config
Already have a token - refresh?
y) Yes
n) No
y/n> y
Use auto config?
 * Say Y if not sure
 * Say N if you are working on a remote or headless machine
y) Yes
n) No
y/n> y
If your browser doesn't open automatically go to the following link: http://127.0.0.1:53682/auth
Log in and authorize rclone for access
Waiting for code...
Got code
Choose a number from below, or type in an existing value
 1 / OneDrive Personal or Business
   \ "onedrive"
 2 / Root Sharepoint site
   \ "sharepoint"
 3 / Type in driveID
   \ "driveid"
 4 / Type in SiteID
   \ "siteid"
 5 / Search a Sharepoint site
   \ "search"
Your choice> 1
Found 1 drives, please select the one you want to use:
0:  (personal) id=XXXXXXXXX
Chose drive to use:> 0
Found drive 'root' of type 'personal', URL: https://onedrive.live.com/?cid=XXXXXXX
Is that okay?
y) Yes
n) No
y/n> y
Current remotes:

Name                 Type
====                 ====
[snip]

e) Edit existing remote
n) New remote
d) Delete remote
r) Rename remote
c) Copy remote
s) Set configuration password
q) Quit config
e/n/d/r/c/s/q> q
```

```
$ rclone lsd onedrive:
2018/09/02 16:32:12 Unable to get drive_id and drive_type. If you are upgrading from older versions of rclone, please run `rclone config` and re-configure this backend.
```

```
$ rclone config show onedrive
--------------------
[onedrive]
type = onedrive
client_id = 
client_secret = 
token = {"access_token":"..."}
--------------------
```

So the `drive_id` and the `drive_type` don't save for some reason.

Can you have a look at this?  It probably isn't the onedrive backend's fault!

**Comment by Cnly** [comment]
The SEGV bug looks like a one posted on the forum before. With this error message in the integration tests it should be easier for us to locate and fix the problem.

**Comment by ncw** [comment]
I put the comment above in #2527 tagged for you and since you've made #2520 I think we are done with this PR :-)

**Pull request closed without merging**